### DISCUSSION TITLE
Adding new tuya device Smart lock TY0A01-Zigbee

### DISCUSSION THREAD

**Discussion by gillesvs** [Koenkk/zigbee2mqtt#11869] [Devices]
Hello,

I am trying to add a tuya smart lock which look like that this one (but in tuya version): 
https://www.zigbee2mqtt.io/devices/A6121.html

I am running zigbee2mqtt in a docker container. I only map the data folder outside the container (not the converters at the moment). But I might have to.

I am following the guide to add a new Tuya device, a smart lock which only exposes the "inserted" message.

I add a generic converter and activate debug.

I get this debug message: 
DEBUG 2022-03-14 16:17:53: RECEIVED ZIGBEE MESSAGE FROM 'SMARTLOCK', TYPE 'ATTRIBUTEREPORT', CLUSTER 'GENBASIC', DATA '{"65506":28,"APPVERSION":88}' FROM ENDPOINT 1 WITH GROUPID 0

So now I have to catch this message which correpond to the insertion of my key in the lock and publish the message....

But I am stuck...

Any idea? 

Thanks

Annexe.
My converter is the one from documentation:

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require("zigbee-herdsman-converters/lib/tuya");

const definition = {
    // Since a lot of Tuya devices use the same modelID, but use different data points
    // it's usually necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TY0A01',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TYST12_3zfaqx1i'
        },
    ],
    model: 'TY0A01-Zigbee',
    vendor: 'Vima',
    description: 'Smart Lock',
    supports: 'inserted',
    fromZigbee: [
        fz.ignore_basic_report, // Add this if you are getting no converter for 'genBasic'
        fz.tuya_data_point_dump, // This is a debug converter, it will be described in the next part
    ],
    toZigbee: [
        tz.tuya_data_point_test, // Another debug converter
    ],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: async (device, coordinatorEndpoint, logger) => {
        const endpoint = device.getEndpoint(1);
        await reporting.bind(endpoint, coordinatorEndpoint, ['genBasic']);
    },
    exposes: [
        // Here you should put all functionality that your device exposes
    ],
};

module.exports = definition;

**Comment by rjulius23** [comment]
Any updates on this ?

**Comment by SandrRu** [comment]
I have same device. 
Smart lock A6121
Detected as:
model 'TY0A01' and manufacturer name '_TYST12_3zfaqx1i'

Any news about support?

**Comment by xX-Nexus-Xx** [comment]
I have as well a zigbee doorlock which is identified as TY0A01 ... but I'm unable to get any information out of the device or to get the instruction set 

Doeas anybody have an idea on how to get this running with Zigbee2mqtt?

thanks for any help

**Comment by MIIB1** [comment]
same problem

any update ?

**Comment by gillesvs** [comment]
Hello,
No sorry, I did not look any further. I think to go to a lock that can actually perform actions and not only give a status.

Regards

**Reply by xX-Nexus-Xx** [reply]
I have done the setup as ZIGBEE device in Tuya and can perform a few actions
- remote unlock the door
- generate a code to unlock the door

My doorlock is closed/locked by default (with auto-lock) so the above functionality helps.

I still would like to control the doorlock with above functions directly via Zigbee2MQTT

I hope somebody had more luck than me to get the codes :(

**Reply by MIIB1** [reply]
Same problem can't make it work.

**Comment by Llenas09** [comment]
Any news? I have Home Assistant with Zigbee2MQTT and get unsupported Device message. I generated the next code in the file TY0A01.js. I think the problem is the device are waiting for a message from gateway because in the Zigbee2MQTT i see the message that the device is correct synchronization but in the "smart lock" i got the message that the pairing is failed. I think that the device doesn't send the correct info until the pairing are completed satisfactory.


With the correct tabulation of the file.
`const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require('zigbee-herdsman-converters/lib/tuya');

const definition = {
    // Since a lot of TuYa devices use the same modelID, but use different datapoints
    // it's necessary to provide a fingerprint instead of a zigbeeModel
    fingerprint: [
        {
            // The model ID from: Device with modelID 'TS0601' is not supported
            // You may need to add \u0000 at the end of the name in some cases
            modelID: 'TY0A01',
            // The manufacturer name from: Device with modelID 'TS0601' is not supported.
            manufacturerName: '_TYST12_bfejtdig',
        },
    ],
    model: 'TY0A01',
    vendor: 'TuYa',
    description: 'Door Lock',
	powerSource: 'Battery',
	supports: 'inserted',
	ota: 'ota.zigbeeOTA',
    fromZigbee: [tuya.fz.datapoints],
    toZigbee: [tuya.tz.datapoints],
    onEvent: tuya.onEventSetTime, // Add this if you are getting no converter for 'commandMcuSyncTime'
    configure: tuya.configureMagicPacket,
    exposes: [
        // Here you should put all functionality that your device exposes
    ],
    meta: {
        // All datapoints go in here
        tuyaDatapoints: [
        ],
    },
};

module. Exports = definition;`

**Comment by Llenas09** [comment]
I hope this helps you, it's a bit strange but I managed to get "lock" and "unlock" messages from the "smart lock" this is fantastic.
I have achieved it using another very similar lock "JAVIS JS-SLK2-ZB". I also get an "action_user" and "action_source" even if I change the registered user or unlock method the numbers don't change so I think it doesn't quite work well. I would like to know where to find the element "fromZigbee: [fz.javis_lock_report" to see if by modifying it I can obtain more information such as the battery or who unlocks the lock. Does anyone know where to find it?

MQTT publish: topic 'zigbee2mqtt/tuya.K-Garaje', payload '{"action":"unlock","action_source":105,"action_user":101,"battery":null,"linkquality":255,"state":"LOCK"}'
MQTT publish: topic 'zigbee2mqtt/tuya.K-Cocina/action', payload 'unlock'


[TY0A01.zip](https://github.com/Koenkk/zigbee2mqtt/files/10178577/TY0A01.zip)

**Reply by thbl** [reply]
> I hope this helps you, it's a bit strange but I managed to get "lock" and "unlock" messages from the "smart lock" this is fantastic. I have achieved it using another very similar lock "JAVIS JS-SLK2-ZB". I also get an "action_user" and "action_source" even if I change the registered user or unlock method the numbers don't change so I think it doesn't quite work well. I would like to know where to find the element "fromZigbee: [fz.javis_lock_report" to see if by modifying it I can obtain more information such as the battery or who unlocks the lock. Does anyone know where to find it?
> 
> MQTT publish: topic 'zigbee2mqtt/tuya.K-Garaje', payload '{"action":"unlock","action_source":105,"action_user":101,"battery":null,"linkquality":255,"state":"LOCK"}' MQTT publish: topic 'zigbee2mqtt/tuya.K-Cocina/action', payload 'unlock'
> 
> [TY0A01.zip](https://github.com/Koenkk/zigbee2mqtt/files/10178577/TY0A01.zip)

did you get luck adding the TY0A01 ?

**Reply by daniperaleda** [reply]
> > I hope this helps you, it's a bit strange but I managed to get "lock" and "unlock" messages from the "smart lock" this is fantastic. I have achieved it using another very similar lock "JAVIS JS-SLK2-ZB". I also get an "action_user" and "action_source" even if I change the registered user or unlock method the numbers don't change so I think it doesn't quite work well. I would like to know where to find the element "fromZigbee: [fz.javis_lock_report" to see if by modifying it I can obtain more information such as the battery or who unlocks the lock. Does anyone know where to find it?
> > MQTT publish: topic 'zigbee2mqtt/tuya.K-Garaje', payload '{"action":"unlock","action_source":105,"action_user":101,"battery":null,"linkquality":255,"state":"LOCK"}' MQTT publish: topic 'zigbee2mqtt/tuya.K-Cocina/action', payload 'unlock'
> > [TY0A01.zip](https://github.com/Koenkk/zigbee2mqtt/files/10178577/TY0A01.zip)
> 
> did you get luck adding the TY0A01 ?

It works but limited as Llenas comment

**Reply by thbl** [reply]
@daniperaleda using the [TY0A01.zip](https://github.com/Koenkk/zigbee2mqtt/files/10178577/TY0A01.zip)  file or ?

**Comment by Hadatko** [comment]
I am considering to buy this. I hope support will be added soon ;)

**Comment by Llenas09** [comment]
This are the DataPoints that I find in the device. But I need help for maching the DataPoints with the zigbee2mqtt.

The part " 'Pendiente', tuya.valueConverter.raw]," is generic.

        tuyaDatapoints: [
			[1, 'Pendiente', tuya.valueConverter.raw], //Fingerprint Unlock
			[2, 'Pendiente', tuya.valueConverter.raw], //Password	
			[3, 'Pendiente', tuya.valueConverter.raw], //Temporary Password
			[4, 'Pendiente', tuya.valueConverter.raw], //Dinamyc Password
			[68, 'Pendiente', tuya.valueConverter.raw], //Unlock by offline password
			[5, 'Pendiente', tuya.valueConverter.raw], //Card
			[8, 'OpenorClose', tuya.valueConverter.onOff], //Open or Close
			[9, 'Alarm', tuya.valueConverter.onOff], //Alarm
            [10, 'battery', tuya.valueConverter.raw], //Battery
            [12, 'Pendiente', tuya.valueConverter.raw], //Reverse Lock
            [14, 'Pendiente', tuya.valueConverter.raw], //Doorbell
            [15, 'Pendiente', tuya.valueConverter.raw], //Choose Ring Tone
            [21, 'Pendiente', tuya.valueConverter.raw], //App unlock within passowrd
            [22, 'Pendiente', tuya.valueConverter.raw], //Answer App
            [23, 'Pendiente', tuya.valueConverter.raw], //Forbidden App Unlock
            [24, 'Pendiente', tuya.valueConverter.raw], //Create Temporary Password
            [25, 'Pendiente', tuya.valueConverter.raw], //Delete Temporary Password
            [26, 'Pendiente', tuya.valueConverter.raw], //Update Temporary Password  
            [27, 'Pendiente', tuya.valueConverter.raw], //Frozen Temporary Password  
            [28, 'Pendiente', tuya.valueConverter.raw], //Unfreeze Temporary Password 
            [35, 'HiJackAlarm', tuya.valueConverter.onOff], //Hijack Alarm
            [39, 'Pendiente', tuya.valueConverter.raw], //Reset Temporary Password  
            [41, 'Pendiente', tuya.valueConverter.raw], //App Unlock
            [47, 'DoorStatus', tuya.valueConverter.onOff], //Door Status 
            [48, 'Pendiente', tuya.valueConverter.raw], //New remote unlock without password - set key
            [49, 'Pendiente', tuya.valueConverter.raw], //New remote unlock without password - safe solution
            [54, 'Pendiente', tuya.valueConverter.raw], //Create unlock methods from app
            [55, 'Pendiente', tuya.valueConverter.raw], //Delete unlock methods from app
            [57, 'LockMotorState', tuya.valueConverter.onOff], //Lock motor state
            [61, 'Pendiente', tuya.valueConverter.raw], //Single use password
            [62, 'Pendiente', tuya.valueConverter.raw], //Remote unlock by voice box

**Reply by xX-Nexus-Xx** [reply]
I was not able to find samples for TUYA datapoints :( ... I really would like to progresss with it because I really like the doorlock

@Llenas09 great work ,, let me know if I can test anything for you. My doorlock is showing 
- random "action" updates (lock/unlock) 
- battery is always NULL
- state is always  ="LOCK:

{
    "last_seen": "2023-05-25T08:06:48+10:00",
    "linkquality": 109,
    "state": "LOCK",
    "action": "",
    "action_source": 105,
    "action_user": 101,
    "battery": null,
    "device": {
        "applicationVersion": 88,
        "dateCode": "",
        "friendlyName": "Entrance.FrontDoor",
        "hardwareVersion": 1,
        "ieeeAddr": "0x540f57fffec91e2e",
        "manufacturerID": 0,
        "manufacturerName": "_TYST12_srptwvak",
        "model": "TY0A01",
        "networkAddress": 62442,
        "powerSource": "Battery",
        "stackVersion": 0,
        "type": "EndDevice",
        "zclVersion": 3
    },
    "update": {
        "installed_version": -1,
        "latest_version": -1,
        "state": null
    },
    "update_available": null
}

**Reply by daniperaleda** [reply]
Thanks to the work of @Llenas09 , I have recovered this

{
    "linkquality": 108,
    "state": "LOCK",
    "action": "",
    "action_source": 105,
    "action_user": 101,
    "battery": null,
    "update": {
        "installed_version": -1,
        "latest_version": -1,
        "state": null
    },
    "update_available": null
}

It is an step, for sure we are all close to get more

**Comment by daniperaleda** [comment]
Using the file TY0A01.zip

Only check the correct manufacturer...in my case "_TYST12_3zfaqx1i"

**Comment by Llenas09** [comment]
Hi, this are the detail of value for DataPoints. I think that the compatibility is closer but I need help for configure DataPoints on the file.

![Tuya-1](https://github.com/Koenkk/zigbee2mqtt/assets/61942524/505cb7f7-7592-4d42-a84e-afcc61064ac4)
![Tuya-2](https://github.com/Koenkk/zigbee2mqtt/assets/61942524/6293f6bf-828b-4ac7-b564-4a8bf300aa08)
![Tuya-3](https://github.com/Koenkk/zigbee2mqtt/assets/61942524/f36d0848-eced-4eb4-b9ee-7281ff1c3aab)
![Tuya-4](https://github.com/Koenkk/zigbee2mqtt/assets/61942524/623a8205-40fb-4a9a-8fc2-284d9b659def)

**Comment by Llenas09** [comment]
Another question for all of you, when I pairing the Smart Lock with Tuya Gateway in the Smart Lock I hear a message that says "Successful". But when are pairing to zigbeetomqtt the device appears in device list with correctly added, but in the SmartLock the lights remain in pairing state until I get a "Fail" message. Is it the same situation for the rest of you?

I think that the SmartLock are waiting a command that says that the pairing are correct. You are know something about that?

**Comment by mmaslo** [comment]
can you please share your exposes variables please

**Comment by Llenas09** [comment]
This is my new file version, but still not working. The values still unknown.

Change the TXT extension to .js
[TY0A01.txt](https://github.com/Koenkk/zigbee2mqtt/files/12816889/TY0A01.txt)

**Reply by mmaslo** [reply]
cheers for that, have you managed to connect successfully?

**Reply by Llenas09** [reply]
No, same status of "Fail" for the pairing. Are you in the same situation?

**Reply by mmaslo** [reply]
yep same, i can see the exposes but wont read any data due to failing the pairing also.

i have emailed Tuya to request all the code for the product, see how i go

**Comment by zeduperes** [comment]
+1 vote I want to use this door lock with zigbee2mqtt.

**Reply by daniperaleda** [reply]
+1 vote. I tried in the past with no response...Hoping we could get it...I am sure that the Zigbee reading is easy...but need to publish the information

**Comment by Night-coyote** [comment]
If i will use Tuya Local integration in HA it will works?

**Comment by Koenkk** [comment]
Let's continue in https://github.com/Koenkk/zigbee2mqtt/issues/4478
